titiler/dependencies.py

```diff
@@ -44,201 +48,201 @@ async def request_hash(request: Request) -> str:
-        self.height = height
-        self.width = width
-        if self.width and self.height:
-            self.max_size = None
-        if nodata is not None:
-            nodata = numpy.nan if nodata == "nan" else float(nodata)
-        self.nodata = nodata
-        self.rescale = rescale
-        self.color_formula = color_formula
-        self.color_map = cmap.get(color_map.value) if color_map else None
-        kwargs = dict(request.query_params)
-        kwargs.pop("url", None)
-        kwargs.pop("format", None)
-        kwargs.pop("bidx", None)
-        kwargs.pop("expression", None)
-        kwargs.pop("nodata", None)
-        kwargs.pop("rescale", None)
-        kwargs.pop("color_formula", None)
-        kwargs.pop("color_map", None)
-        kwargs.pop("height", None)
-        kwargs.pop("width", None)
-        kwargs.pop("max_size", None)
-        kwargs.pop("assets", None)  # For STAC
-        self.kwargs = kwargs
-
-
-class CommonMetadataParams:
+def MosaicTMSParams(
+    TileMatrixSetId: MosaicTileMatrixSetNames = Query(
+        MosaicTileMatrixSetNames.WebMercatorQuad,  # type: ignore
+        description="TileMatrixSet Name (default: 'WebMercatorQuad')",
+    )
+) -> morecantile.TileMatrixSet:
+    """TileMatrixSet Dependency."""
+    return morecantile.tms.get(TileMatrixSetId.name)
+
+
+@dataclass
+class DefaultDependency:
+    """Dependency Base Class"""
+
+    kwargs: dict = field(init=False, default_factory=dict)
+
+
+@dataclass
+class PathParams(DefaultDependency):
+    """Create dataset path from args"""
+
+    url: str = Query(..., description="Dataset URL")
+    reader: Optional[Type[BaseReader]] = field(init=False, default=None)  # Placeholder
+
+    def __post_init__(self,):
+        """Define dataset URL."""
+        parsed = urlparse(self.url)
+
+        # Placeholder in case we want to allow landsat+mosaicid
+        # by default we store the mosaicjson as a GZ compressed json (.json.gz) file
+        if parsed.scheme.split("+")[-1] == "mosaicid":
+            self.url = f"{settings.DEFAULT_MOSAIC_BACKEND}{settings.DEFAULT_MOSAIC_HOST}/{parsed.netloc}.json.gz"
+
+        # if parsed.scheme.startswith("landsat+"):
+        #     self.reader = L8Reader
+        #     self.url = self.url.replace("landsat+", "")
+        # elif parsed.scheme.startswith("sentinel2cog+"):
+        #     self.reader = Sentinel2COG
+        #     self.url = self.url.replace("sentinel2cog+", "")
+        # ....
+
+
+@dataclass
+class AssetsParams(DefaultDependency):
+    """Create dataset path from args"""
+
+    assets: Optional[str] = Query(
+        None,
+        title="Asset indexes",
+        description="comma (',') delimited asset names (might not be an available options of some readers)",
+    )
+    kwargs: dict = field(init=False, default_factory=dict)
+
+    def __post_init__(self):
+        """Post Init."""
+        if self.assets is not None:
+            self.kwargs["assets"] = self.assets.split(",")
```

Comment:
⚠️ we use kwargs to store the `assets` option so we can just pass `**params.kwargs` to the function (in the tiler) and we do not have to worry if the reader accept the asset keyword. 

in the tiler factory with use `add_asset_deps` to specify if we use `AssetsParams` or `DefaultDependency`. The deps will populate the open api with assets (if set to true) but on the function we just need to declare 

```python
        optionsParams = AssetsParams if self.add_asset_deps else DefaultDependency
        
        def info(
            src_path=Depends(self.path_dependency), 
            options: optionsParams = Depends()
        ):
            """Return basic info."""
            with reader(src_path.url) as src_dst:
                info = src_dst.info(**options.kwargs)

```